File: problems/74-Search-a-2D-Matrix/solution.go
func searchMatrix(matrix [][]int, target int) bool {
	row, col := len(matrix), len(matrix[0])
	midRow, midCol := row/2, 0

	if target < matrix[midRow][0] {
		if midRow > 0 {
			return searchMatrix(matrix[:midRow], target)
		} else {
			return false
		}
	} else if target > matrix[midRow][col-1] {
		if midRow < row-1 {
			return searchMatrix(matrix[midRow+1:], target)
		} else {
			return false
		}
	} else {
		start, end := 0, col-1
		for start < end {
			midCol = (start + end) / 2
			if matrix[midRow][midCol] == target {
				return true
			} else if matrix[midRow][midCol] > target {
				end = midCol
			} else {
				start = midCol + 1
			}
		}

		midCol = (start + end) / 2
		if matrix[midRow][midCol] == target {
			return true
		} else {
			return false
		}
	}
}

File: problems/74-Search-a-2D-Matrix/solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rS = 0, rE = matrix.length - 1;
        while (rS < rE) {
            int rH = (rS + rE) / 2;

            if (target <= matrix[rH][0]) {
                rE = rH;
            } else {
                rS = rH + 1;
            }
        }

        if (target < matrix[rS][0]) {
            rS--;
        }

        if (rS < 0) {
            return false;
        }

        int cS = 0, cE = matrix[rS].length - 1;
        while (cS < cE) {
            int cH = (cS + cE) / 2;

            if (target <= matrix[rS][cH]) {
                cE = cH;
            } else {
                cS = cH + 1;
            }
        }

        if (target != matrix[rS][cS]) {
            return false;
        } else {
            return true;
        }
    }
}

